Move @Slynova/flydrive to local package + add support for Azure Blob Storage

TheAzack9 opened the pull request:
As discussed with @rijkvanzanten on Discord does this PR move the flydrive package to a local folder and extends the package to include support for Azure blob storage. This is great for everyone living in the azure side of the cloud as it allows us to store files in the azure cloud. I have tried to update all documentation that i could find to show that azure blobs are included :)

I did strip tests from the flydrive package as i'm not sure if those are relevant for Directus or not. (I can add them back if requested) I am not sure what is relevant or not to keep exactly, so please just say if there are files i should remove or shouldn't have removed :) (I have not removed anything essential to the packages ofc)

I have not done extensive testing on this so far, but i surely will going forward as this is something i need myself ;) 

I hope it looks good 👍

rijkvanzanten commented on package.json:
```diff
@@ -189,7 +189,11 @@
     "create-directus-project": "file:packages/create-directus-project",
-    "directus": "file:api"
+    "directus": "file:api",
+    "@slynova/flydrive": "file:packages/drive/packages/flydrive",
+    "@slynova/flydrive-gcs": "file:packages/drive/packages/flydrive-gcs",
+    "@slynova/flydrive-s3": "file:packages/drive/packages/flydrive-s3",
+    "@slynova/flydrive-azureBlob": "file:packages/drive/packages/flydrive-azureBlob"
```
Change this to `@directus/drive`

rijkvanzanten commented on api/src/storage.ts:
```diff
@@ -3,15 +3,16 @@ import {
 	StorageManagerConfig,
 	Storage,
-} from '@slynova/flydrive';
+} from '@directus/flydrive';
```
Lets name it `@directus/drive` instead

rijkvanzanten commented on api/src/services/assets.ts:
```diff
@@ -44,17 +44,17 @@ export class AssetsService {
 
 			if (exists) {
-				return { stream: storage.disk(file.storage).getStream(assetFilename), file };
+				return { stream: await storage.disk(file.storage).getStream(assetFilename), file };
```
Did we document this change in the readme?

rijkvanzanten commented on api/package.json:
```diff
@@ -69,9 +69,10 @@
+		"@directus/flydrive": "file:../packages/drive/packages/flydrive",
+		"@directus/flydrive-gcs": "file:../packages/drive/packages/flydrive-gcs",
+		"@directus/flydrive-s3": "file:../packages/drive/packages/flydrive-s3",
+		"@directus/flydrive-azureBlob": "file:../packages/drive/packages/flydrive-azureBlob",
```
Use lowercase for the package name `azureblob` or `azure-blob`, update references

rijkvanzanten replied:
We'll have to "lift" these sub-packages into the "root" packages folder:

`/packages/drive/packages/flydrive-azureBlob` -> `/packages/drive-azure-blob`

rijkvanzanten reviewed (commented):
Instead of `azure-blob` (and potentially later `azure-file`), would it be possible to have a single `drive-azure`, and support both blobs & files based on a config flag?

rijkvanzanten commented on packages/drive/.gitignore:
```diff
@@ -0,0 +1,6 @@
+node_modules
```
Lets use the same gitignore / eslint / editorconfig setup as the other packages

rijkvanzanten commented on packages/drive/README.md:
```diff
@@ -0,0 +1,314 @@
+<p align="center">
```
We'll have to clean this up / rewrite to make sense

rijkvanzanten commented on packages/drive/lerna.json:
```diff
@@ -0,0 +1,7 @@
+{
```
We'll have to get rid of this, can't have nested lerna configurations

rijkvanzanten commented on packages/drive/package.json:
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "root",
```
Needs to go in favor of being `@directus/drive` "root"

rijkvanzanten commented on packages/drive/packages/flydrive-azureBlob/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+{
```
Use root directus tsconfig

rijkvanzanten reviewed (changes requested):
Don't take these points as hard to-do's @TheAzack9 🙂 I can pick up where you left off 👍

TheAzack9 commented:
> Instead of `azure-blob` (and potentially later `azure-file`), would it be possible to have a single `drive-azure`, and support both blobs & files based on a config flag?

This should be fine to do (and probably is the best solution tbh, easy to get tunnel visioned ;) )

> Don't take these points as hard to-do's @TheAzack9 🙂 I can pick up where you left off 👍

No worries, i just want to help get the ball rolling as much as i can ;) I can gladly go through tomorrow and fix what i can. I kinda want to play around some more with it, so should be fun! :)

TheAzack9 commented:
Took a bit longer than planned, but i finally found the time to go through and fix the comments :)

I have tried to change stuff to be more consistent with existing configs, but there may still be some stuff i have misplaced 😅 
Short summary of changes:

- Renamed from "slynova/flydrive*" -> "directus/drive*"
- Removed learna config in sub directory. I still kept the packages in a subdirectory (/packages/drive) since it looked more organized imo and has better separation since they are related. Should be simple to move them even further if necessary. (Also allows for a general readme, which i feel is nice wrt. giving credit and such. 
- Documented the async change to getStream
- Renamed from `directus/azureBlob` to `directus/azure`
- Removed a lot of uncecessary configs that is duplicate to root

Remaining tasks:

- Update readme in packages/drive/ :)

emregency commented:
Hey guys,

I was just wondering what is actually blocking this PR and how we can move it forward. I am looking forward to use Azure Blob so if there anything I can do to move this PR further let me know. It looks like great work was put in and needs just a last push.

benhaynes commented:
Time. Our team has been extremely busy wrapping up other internal systems, and will _finally_ be back to the Core codebase this week. I defer to @rijkvanzanten on what can be done (if anything) to expedite.

TheAzack9 commented:
Just jumping in with a little ping! :) 

If there is any more i could do to help here i would love to help out 😅 I am depending a bit on this to be able to go live with a website :)

rijkvanzanten commented:
@TheAzack9 Yes actually! I just recently dove into this to address #3640 and was wondering about 2 things: 

Does getStream need to be async, and

can you maybe add in support for ranges? :o

I believe every storage driver we have has native support for ranged requests, which in turn allows us to expose that on the /assets endpoint

Other than that, there's no blockers

TheAzack9 commented:
I thought it had to, but i see that Xalag proposed a solution that is not breaking here: https://github.com/Slynova-Org/flydrive/pull/158, so it should be possible for it to not be async :)

I can look into supporting ranges :)

TheAzack9 commented:
I need some more context on what you mean with ranged requests i think @rijkvanzanten :) Are you thinking of this? https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests

rijkvanzanten commented:
Yes exactly! It's a way to ask for specific bytes out of a file. Browsers use this for things like video buffering, where they only requests little parts of a video at a time. 

From a quick search around, it seems like all the external libraries used support it in some way or fashion. This is how you do it in "local" (eg FS) for example:

<img width="1022" alt="CleanShot 2021-02-04 at 13 37 12@2x" src="https://user-images.githubusercontent.com/9141017/106939349-268daa00-66ee-11eb-930a-0a019c92c65f.png">
